subxt/src/events/events_client.rs

```diff
@@ -73,7 +85,7 @@ where
                 .map(|e| e.0)
                 .unwrap_or_else(Vec::new);
 
-            Ok(Events::new(client.metadata(), block_hash, event_bytes))
+            Ok(Events::new(metadata, block_hash, event_bytes))
```

Comment:
Maybe the solution really is that we allow the user to manually do the above and obtain metadata for the older block if they are in doubt, and ensure that they can construct this `Events` object.

It might be nicer then to move the `event_bytes` stuff into `Events::new` and make that async or something, so that the interface is simpler (they provide metadata and block hash and get events back for it).

Reply:
I recall that comparing metadata is super expensive as well, maybe there is some simple way to find which runtime version a block belongs to without using some block explorer API. So how would I know if a block is "old" i.e. need to call `metadata`? I suppose one could do some simple calculation based on that 14 400 blocks per day is produced * 30 or something.

My point is that comparing metadata is superslow as well then fetching metadata could make sense to just be naive and always fetch the metadata regardless.....  so folks doesn't end up with decoding errors.

However, I agree it's probably better to have different APIs or take `Option<Metadata>` as input in `fn at` i.e, if None just use the latest.